linux/shim/init.go

```diff
@@ -60,15 +62,27 @@ func newInitProcess(context context.Context, path, namespace string, r *shimapi.
 		}
 		options = *v.(*runcopts.CreateOptions)
 	}
+
+	rootfs := filepath.Join(path, "rootfs")
+	// count the number of successful mounts so we can undo
+	// what was actually done rather than what should have been
+	// done.
+	nrRootMounts := 0
 	for _, rm := range r.Rootfs {
 		m := &mount.Mount{
 			Type:    rm.Type,
 			Source:  rm.Source,
 			Options: rm.Options,
 		}
-		if err := m.Mount(filepath.Join(path, "rootfs")); err != nil {
+		if err := m.Mount(rootfs); err != nil {
 			return nil, errors.Wrapf(err, "failed to mount rootfs component %v", m)
 		}
+		nrRootMounts++
+	}
+	cleanupMounts := func() {
+		if err2 := mount.UnmountN(rootfs, 0, nrRootMounts); err2 != nil {
+			log.G(context).WithError(err2).Warn("Failed to cleanup rootfs mount")
+		}
```

Comment:
The `err` captured by the closure may not be the same `err` as gets assigned to further down the function, since `err` can be shadowed (e.g. `if err := foo(); err != nil { return err }` shadows the previous `err`, I think)